Add RotateAbsolute action to turtlesim

jacobperron opened the pull request:
Resolves #61

jacobperron committed:
[turtlesim] Add action interface for rotating to absolute angle

Signed-off-by: Jacob Perron <jacob@openrobotics.org>

jacobperron committed:
Implement RotateAbsolute action

Only one action is executed at a time. Any new actions override existing ones.
If a velocity command is received while executing the action, then the goal is aborted.

Signed-off-by: Jacob Perron <jacob@openrobotics.org>

dirk-thomas commented:
In order to trigger this action easily it would be good to update keyboard teleop with a few key to rotate to e.g. orientations in 90 (e.g. R|G|V|D) / 45 (e.g. R|T|G|B|V|C|D|E) degrees steps.

jacobperron commented:
> In order to trigger this action easily it would be good to update keyboard teleop with a few key to rotate to e.g. orientations in 90 (e.g. R|G|V|D) / 45 (e.g. R|T|G|B|V|C|D|E) degrees steps.

I was just about to post some commands for testing this. Since this is intended to go along with the tutorial, I figured the user will be prompted to send the actions with `ros2 action` tool. For example:

    ros2 action send_goal /turtle1/rotate_absolute turtlesim/action/RotateAbsolute "theta: 1.57"

another with feedback:

    ros2 action send_goal /turtle1/rotate_absolute turtlesim/action/RotateAbsolute "theta: 1.0" --feedback

Adding to the keyboard teleop isn't a bad idea either. Want me to follow up in a second PR?

dirk-thomas commented:
> Want me to follow up in a second PR?

Either a separate one or added to this. I don't mind either way.

jacobperron committed:
Support triggering rotation action with keyboard teleop

Use the keys g/b/v/c/d/e/r/t to rotate the turtle to absolute orientations and 'f' to cancel a rotation.

Signed-off-by: Jacob Perron <jacob@openrobotics.org>

jacobperron commented:
@dirk-thomas See 0e6dec2 for keyboard teleop support. I also made the 'F' key cancel an action.

jacobperron committed:
Catch exceptions when attempting to cancel goal

Signed-off-by: Jacob Perron <jacob@openrobotics.org>

dirk-thomas commented on turtlesim/src/turtle.cpp:
```diff
@@ -34,13 +34,20 @@
 #include <QColor>
 #include <QRgb>
 
+#define TWO_PI 6.283185307179586
```
Why not use the definition of `PI` from `turtlesim/turtle.h`?

jacobperron replied:
See 5f614f0

dirk-thomas commented:
With this patch when quiting the teleop node the console is "unusable" since the state isn't correctly restored anymore. E.g. you don't see what you type. Since that is not the case on the target branch it seems to be related to the changes.

jacobperron commented:
> With this patch when quiting the teleop node the console is "unusable" since the state isn't correctly restored anymore. E.g. you don't see what you type. Since that is not the case on the target branch it seems to be related to the changes.

Odd. It doesn't happen for me in ZSH, but I can reproduce the issue with Bash...

jacobperron committed:
Add back input.shutdown()

Signed-off-by: Jacob Perron <jacob@openrobotics.org>

jacobperron commented:
@dirk-thomas  c31fab3 seems to resolve the shell issue after shutdown.

dirk-thomas commented on turtlesim/tutorials/teleop_turtle_key.cpp:
```diff
@@ -145,20 +243,30 @@ int main(int argc, char** argv)
 
   int rc = teleop_turtle.keyLoop();
   input.shutdown();
-  rclcpp::shutdown();
```
Why is this being removed?

jacobperron replied:
It shouldn't have been, my mistake. Restored 55c5aa4

jacobperron committed:
Add back rclcpp::shutdown

Signed-off-by: Jacob Perron <jacob@openrobotics.org>

jacobperron committed:
Define TWO_PI from PI in header

Signed-off-by: Jacob Perron <jacob@openrobotics.org>

dirk-thomas committed:
match style

dirk-thomas commented:
Thanks for the new functionality. Looks good to me :+1:

dirk-thomas commented on turtlesim/tutorials/teleop_turtle_key.cpp:
```diff
@@ -55,29 +67,74 @@ class KeyboardReader
         }
         else if (buffer.Event.KeyEvent.wVirtualKeyCode == VK_DOWN)
+        {
+          *c = KEYCODE_DDOWN;
```
`undeclared identifier`: see #66 for a fix.

dirk-thomas commented on turtlesim/tutorials/teleop_turtle_key.cpp:
```diff
@@ -123,6 +184,43 @@ TeleopTurtle::TeleopTurtle():
+void TeleopTurtle::sendGoal(double theta)
+{
+  auto goal = turtlesim::action::RotateAbsolute::Goal();
+  goal.theta = theta;
```
@jacobperron https://ci.ros2.org/job/ci_windows/8159/warnings43Result/new/

> '=': conversion from 'double' to 'turtlesim::action::RotateAbsolute_Goal_>::_theta_type', possible loss of data

jacobperron replied:
See https://github.com/ros/ros_tutorials/pull/69 for a fix
